tortoisebeacon/tortoise_beacon.go

```diff
@@ -503,84 +503,61 @@ func (tb *TortoiseBeacon) runConsensusPhase(ctx context.Context, epoch types.Epo
 	// For next rounds,
 	// wait for δ time, and construct a message that points to all messages from previous round received by δ.
 	// rounds 1 to K
-	timer := time.NewTimer(tb.config.FirstVotingRoundDuration + tb.config.WeakCoinRoundDuration)
+	timer := time.NewTimer(tb.config.FirstVotingRoundDuration)
 	defer timer.Stop()
 
 	var (
-		coinFlip            bool
-		ownLastRoundVotesMu sync.RWMutex
-		ownLastRoundVotes   allVotes
+		ownVotes allVotes
+		err      error
 	)
-
 	for round := firstRound; round <= tb.lastRound(); round++ {
-		// always use coinflip from the previous round for current round.
-		// round 1 is running without coinflip (e.g. value is false) intentionally
 		round := round
-		previousCoinFlip := coinFlip
+		votes := ownVotes
 		tb.eg.Go(func() error {
 			if round == firstRound {
 				if err := tb.sendProposalVote(ctx, epoch); err != nil {
-					logger.With().Error("Failed to send proposal vote",
+					logger.With().Error("failed to send proposal vote",
+						log.Uint32("round_id", uint32(round)),
+						log.Err(err))
+				}
+			} else {
+				if err := tb.sendFollowingVote(ctx, epoch, round, votes); err != nil {
+					logger.With().Error("failed to send following vote",
 						log.Uint32("round_id", uint32(round)),
 						log.Err(err))
-
-					return fmt.Errorf("failed to send proposal vote: %w", err)
 				}
-
-				return nil
-			}
-
-			// next rounds, send vote
-			// construct a message that points to all messages from previous round received by δ
-			ownCurrentRoundVotes, err := tb.calcVotes(epoch, round, previousCoinFlip)
-			if err != nil {
-				logger.With().Error("Failed to calculate votes",
-					log.Uint32("round_id", uint32(round)),
-					log.Err(err))
-
-				return fmt.Errorf("calculate votes: %w", err)
-			}
-
-			if round == tb.lastRound() {
-				ownLastRoundVotesMu.Lock()
-				ownLastRoundVotes = ownCurrentRoundVotes
-				ownLastRoundVotesMu.Unlock()
-			}
-
-			if err := tb.sendFollowingVote(ctx, epoch, round, ownCurrentRoundVotes); err != nil {
-				logger.With().Error("Failed to send following vote",
-					log.Uint32("round_id", uint32(round)),
-					log.Err(err))
-
-				return fmt.Errorf("send following vote: %w", err)
 			}
-
-			return nil
-		})
-
-		tb.eg.Go(func() error {
-			tb.startWeakCoinRound(ctx, epoch, round)
 			return nil
 		})
+		if round != firstRound {
+			tb.eg.Go(func() error {
+				tb.startWeakCoinRound(ctx, epoch, round)
+				return nil
+			})
+		}
 
 		select {
 		case <-timer.C:
 			timer.Reset(tb.config.VotingRoundDuration + tb.config.WeakCoinRoundDuration)
```

Comment:
this needs to adjust too.

are we sure in the test params that voting round duration is a lot longer than weakcoin duration?

Reply:
> this needs to adjust too.

you mean if votes will have to be tallied twice?

> are we sure in the test params that voting round duration is a lot longer than weakcoin duration?

it is longer by 10-15s. why do you think that it should be *a lot* longer?

Reply:
oh. i just mean that in this order of operation, VotingRoundDuration includes weakCoinRoundDuration. 
conceptually a voting round includes sending votes, generating weak coin, and counting votes. so i thought VotingRoundDuration covers all those.